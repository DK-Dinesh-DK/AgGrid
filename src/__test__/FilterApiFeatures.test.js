import { render, act, screen, fireEvent } from "@testing-library/react";
import DataGrid from "../components/datagrid/DataGrid";
import { useRef } from "react";
import "@testing-library/jest-dom";
let getRenderedrowsOutput;
let getFilterModelOutput;
const columns = [
  {
    field: "id",
    headerName: "ID",
    width: 80,
  },
  {
    field: "task",
    headerName: "Title",
  },
  {
    field: "priority",
    headerName: "Priority",
  },
];
const rows = [
  { id: 1, task: "Task #1", priority: "Low" },
  { id: 2, task: "Task #2", priority: "Medium" },
  { id: 3, task: "Task #3", priority: "High" },
  { id: 4, task: "Task #4", priority: "Medium" },
  { id: 5, task: "Task #5", priority: "Low" },
  { id: 6, task: "Task #6", priority: "High" },
  { id: 7, task: "Task #7", priority: "Critical" },
];
function LaiDatagrid(props) {
  const dataGridRef = useRef(null);
  return (
    <>
      <button
        type="button"
        data-testid="setQuickFilterBtn"
        onClick={() =>
          dataGridRef.current.api.setQuickFilter(props.filterString)
        }
      >
        setQuickFilter
      </button>
      <button
        type="button"
        data-testid="getRenderedrowsBtn"
        onClick={() =>
          (getRenderedrowsOutput = dataGridRef.current.api.getRenderedrows())
        }
      >
        getRenderedrows
      </button>
      <button
        type="button"
        data-testid="getFilterModelBtn"
        onClick={() =>
          (getFilterModelOutput = dataGridRef.current.api.getFilterModel())
        }
      >
        getFilterModel
      </button>
      <button
        type="button"
        data-testid="setFilterModelBtn"
        onClick={() =>
          dataGridRef.current.api.setFilterModel(props.filterModel)
        }
      >
        setFilterModel
      </button>
      <button
        type="button"
        data-testid="isColumnFilterPresent"
        onClick={() => dataGridRef.current.api.isColumnFilterPresent()}
      >
        isColumnFilterPresent
      </button>
      <button
        type="button"
        data-testid="isAnyFilterPresent"
        onClick={() => dataGridRef.current.api.isAnyFilterPresent()}
      >
        isAnyFilterPresent
      </button>
      <button
        type="button"
        data-testid="setColumnDefs"
        onClick={() =>
          dataGridRef.current.api.setColumnDefs([
            ...columns,
            { headerName: "New Column", field: "newColumn", filter: true },
          ])
        }
      >
        setColumnDefs
      </button>

      <DataGrid
        columnData={columns}
        rowData={rows}
        enableVirtualization={false}
        innerRef={dataGridRef}
        testId={"laidatagrid"}
      />
    </>
  );
}
describe("GridAPI - Filtering APIs", () => {
  test("setQuickFilter works correctly", () => {
    const filterString = "high";
    const expectedRowsAfterFilter = [
      { id: 3, task: "Task #3", priority: "High" },
      { id: 6, task: "Task #6", priority: "High" },
    ];
    render(<LaiDatagrid filterString={filterString} />);

    const datagrid = screen.queryByTestId("laidatagrid");
    expect(datagrid).toBeInTheDocument();

    const getRenderedrowsBtn = screen.queryByTestId("getRenderedrowsBtn");
    expect(getRenderedrowsBtn).toBeInTheDocument();

    act(() => fireEvent.click(getRenderedrowsBtn));

    expect(getRenderedrowsOutput).toEqual(rows);

    const setQuickFilterBtn = screen.queryByTestId("setQuickFilterBtn");
    expect(setQuickFilterBtn).toBeInTheDocument();

    act(() => {
      fireEvent.click(setQuickFilterBtn);
    });

    act(() => fireEvent.click(getRenderedrowsBtn));

    expect(getRenderedrowsOutput).toEqual(expectedRowsAfterFilter);
  });
  test("getFilterModel,setFilterModel work correctly", () => {
    const filterString = "high";
    const initialFilterModel = {
      id: "",
      task: "",
      priority: "",
      undefined: "",
      enabled: true,
    };
    const expectedFilterModel = {
      id: "",
      task: "Task #2",
      priority: "",
      undefined: "",
      enabled: true,
    };
    render(
      <LaiDatagrid
        filterString={filterString}
        filterModel={{ task: "Task #2" }}
      />
    );

    const datagrid = screen.queryByTestId("laidatagrid");
    expect(datagrid).toBeInTheDocument();

    const getRenderedrowsBtn = screen.queryByTestId("getRenderedrowsBtn");
    expect(getRenderedrowsBtn).toBeInTheDocument();

    act(() => fireEvent.click(getRenderedrowsBtn));

    expect(getRenderedrowsOutput).toEqual(rows);

    const getFilterModelBtn = screen.queryByTestId("getFilterModelBtn");
    expect(getFilterModelBtn).toBeInTheDocument();

    act(() => {
      fireEvent.click(getFilterModelBtn);
    });

    expect(getFilterModelOutput).toEqual(initialFilterModel);

    const setFilterModelBtn = screen.queryByTestId("setFilterModelBtn");
    expect(setFilterModelBtn).toBeInTheDocument();

    act(() => {
      fireEvent.click(setFilterModelBtn);
    });
    act(() => {
      fireEvent.click(getRenderedrowsBtn);
      fireEvent.click(getFilterModelBtn);
    });

    expect(getRenderedrowsOutput).toEqual([...rows[1]]);
    expect(getFilterModelOutput).toEqual(expectedFilterModel);
  });
  test("isColumnFilterPresent", () => {
    render(<LaiDatagrid filterModel={{ task: "Task #2" }} />);

    const datagrid = screen.queryByTestId("laidatagrid");
    expect(datagrid).toBeInTheDocument();

    const isColumnFilterPresentBtn = screen.queryByTestId(
      "isColumnFilterPresent"
    );
    expect(isColumnFilterPresentBtn).toBeInTheDocument();

    act(() => fireEvent.click(isColumnFilterPresentBtn));
    const setFilterModelBtn = screen.queryByTestId("setFilterModelBtn");
    expect(setFilterModelBtn).toBeInTheDocument();

    act(() => {
      fireEvent.click(setFilterModelBtn);
    });
    act(() => fireEvent.click(isColumnFilterPresentBtn));
  });
  test("isAnyFilterPresent", () => {
    const filterString = "high";
    render(
      <LaiDatagrid
        filterString={filterString}
        filterModel={{ task: "Task #2" }}
      />
    );

    const datagrid = screen.queryByTestId("laidatagrid");
    expect(datagrid).toBeInTheDocument();

    const isAnyFilterPresentBtn = screen.getByTestId("isAnyFilterPresent");
    expect(isAnyFilterPresentBtn).toBeInTheDocument();
    act(() => fireEvent.click(isAnyFilterPresentBtn));

    const setColumnDefsBtn = screen.queryByTestId("setColumnDefs");
    expect(setColumnDefsBtn).toBeInTheDocument();
    act(() => {
      fireEvent.click(setColumnDefsBtn);
    });
    act(() => fireEvent.click(isAnyFilterPresentBtn));
  });
});
