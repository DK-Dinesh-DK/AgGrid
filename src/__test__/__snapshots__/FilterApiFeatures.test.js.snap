// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridAPI - Filtering APIs selectAllFiltered and deselectAllFiltered work correctly 1`] = `
<div
  aria-colcount="3"
  aria-rowcount="3"
  class="rdg r1fgqq0p"
  data-testid="laidatagrid"
  dir="ltr"
  id="DataGrid"
  role="grid"
  style="background: rgb(184, 204, 228); grid-template-rows: 24px repeat(2, 24px); --rdg-header-row-height: 24px; --rdg-summary-row-height: 24px; --rdg-sign: 1; grid-template-columns: 80px 40px 40px;"
>
  <div
    aria-rowindex="1"
    class="rdg-header-row h1w9imsy"
    role="header-row"
    style="--rdg-grid-row-start: 1;"
  >
    <div
      aria-colindex="1"
      class="rdg-cell c9tygie rdg-header-column-even"
      draggable="true"
      role="parentcolumn"
      style="height: 100%; grid-column-start: 1;"
      tabindex="0"
    >
      <div
        style="display: flex; height: 100%; align-items: center; justify-content: center;"
      >
        <div
          aria-selected="false"
          role="columnheader"
          style="height: 24px; display: flex; justify-content: center; width: 100%;"
        >
          <div
            style="display: flex; justify-content: center; align-items: center; height: inherit; width: 80px; padding-left: 5px; padding-right: 5px;"
          >
            ID
          </div>
        </div>
      </div>
    </div>
    <div
      aria-colindex="2"
      class="rdg-cell c9tygie rdg-header-column-odd"
      draggable="true"
      role="parentcolumn"
      style="height: 100%; grid-column-start: 2;"
      tabindex="-1"
    >
      <div
        style="display: flex; height: 100%; align-items: center; justify-content: center;"
      >
        <div
          aria-selected="false"
          role="columnheader"
          style="height: 24px; display: flex; justify-content: center; width: 100%;"
        >
          <div
            style="display: flex; justify-content: center; align-items: center; height: inherit; width: auto; padding-left: 5px; padding-right: 5px;"
          >
            Title
          </div>
        </div>
      </div>
    </div>
    <div
      aria-colindex="3"
      class="rdg-cell c9tygie rdg-header-column-even"
      draggable="true"
      role="parentcolumn"
      style="height: 100%; grid-column-start: 3;"
      tabindex="-1"
    >
      <div
        style="display: flex; height: 100%; align-items: center; justify-content: center;"
      >
        <div
          aria-selected="false"
          role="columnheader"
          style="height: 24px; display: flex; justify-content: center; width: 100%;"
        >
          <div
            style="display: flex; justify-content: center; align-items: center; height: inherit; width: auto; padding-left: 5px; padding-right: 5px;"
          >
            Priority
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-rowindex="2"
    class="rdg-row rsqlkts rdg-row-even"
    id="3"
    role="row"
    style="--rdg-grid-row-start: 2; --rdg-row-height: 24px;"
  >
    <div
      aria-colindex="1"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-even"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 1; grid-row-end: span [object Object]; --rdg-summary-row-top: 24px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        3
      </div>
    </div>
    <div
      aria-colindex="2"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-odd"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 2; grid-row-end: span [object Object]; --rdg-summary-row-top: 24px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        Task #3
      </div>
    </div>
    <div
      aria-colindex="3"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-even"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 3; grid-row-end: span [object Object]; --rdg-summary-row-top: 24px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        High
      </div>
    </div>
  </div>
  <div
    aria-rowindex="3"
    class="rdg-row rsqlkts rdg-row-odd"
    id="6"
    role="row"
    style="--rdg-grid-row-start: 3; --rdg-row-height: 24px;"
  >
    <div
      aria-colindex="1"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-even"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 1; grid-row-end: span [object Object]; --rdg-summary-row-top: 48px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        6
      </div>
    </div>
    <div
      aria-colindex="2"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-odd"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 2; grid-row-end: span [object Object]; --rdg-summary-row-top: 48px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        Task #6
      </div>
    </div>
    <div
      aria-colindex="3"
      aria-readonly="true"
      class="rdg-cell c9tygie rdg-cell-column-even"
      data-testid="rowCell"
      style="height: 100%; grid-column-start: 3; grid-row-end: span [object Object]; --rdg-summary-row-top: 48px;"
      tabindex="-1"
    >
      <div
        aria-selected="false"
        role="gridcell"
        style="width: 100%; text-align: center; text-overflow: ellipsis; overflow: hidden; height: inherit; padding-inline: 6px;"
      >
        High
      </div>
    </div>
  </div>
  <div
    class="m22dhxs"
    data-measuring-cell-key="id"
    style="grid-column-start: 1; min-width: 40px;"
  />
  <div
    class="m22dhxs"
    data-measuring-cell-key="task"
    style="grid-column-start: 2; min-width: 40px;"
  />
  <div
    class="m22dhxs"
    data-measuring-cell-key="priority"
    style="grid-column-start: 3; min-width: 40px;"
  />
</div>
`;
